### PR TITLE
Enumerable Methods

### PR DESCRIPTION
**Pull request by okikiola11**
- Added the #my_each method using the yield statement
- Added the #my_each_with_index method using the yield statement.
- Added #my_select using the yield statement
- Added #my_any? method
- Added  #my_none? method 
- Added #my_count method
- Added #my_map method
- Added #my_inject method

### CODE REVIEW

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    result = []
+    i = 0
+    while i < size
+      result << self[i] if yield(self[i]) == true
+      i += 1
+    end
+  end
```
Comment (Torres-ssf) [inline enumerable_methods.rb]: my_select...

- my_select should return an array containing all elements of enum for which the given block returns a true value.

`[1,2,3,4,5].select { |num|  num.even?  }` returns  [2, 4]
`a.select { |v| v =~ /[aeiou]/ }`  returns ["a", "e"]

Both of these tests on your code returns nil.

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      my_each { |item| condition = false if item.is_a?(args) }
+    elsif args.is_a?(Regexp)
+      my_each { |item| condition = false if args.match?(item.to_s) }
+    elsif !args.nil?
+      my_each { |item| condition = false if item != args }
+    else
+      my_each { |item| condition = false if item }
+    end
+    condition
+  end
```
Comment (Torres-ssf) [inline enumerable_methods.rb]: my_all?

- when no block or argument is given returns true when none of the collection members are false or nil

`[nil, true, 99].all?` return false
`[2, 5, 'a'].all?` return true   ===>  **your code returns false here**



- when a class is passed as an argument returns true if all of the collection is a member of such class

`[1, 2i, 3.14].all?(Numeric)` returns true



- when a Regex is passed as an argument returns true if all of the collection matches the Regex

`%w[ant bear cat].all?(/t/)`  returns false
`%w[ant bit boot].all?(/t/)` returns true

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      my_each { |item| condition = true if item.is_a?(args) }
+    elsif args.is_a?(Regexp)
+      my_each { |item| condition = true if args.match?(item.to_s) }
+    elsif args.nil?
+      my_each { |item| condition = true if item }
+    else
+      my_each { |item| condition = true if item == args }
+    end
+    condition
+  end
```
Comment (Torres-ssf) [inline enumerable_methods.rb]: my_any? returns true if the block ever returns a value other than false or nil

`%w[ant bear cat].any? { |word| word.length >= 3 }`  returns true
`%w[ant bear cat].any? { |word| word.length > 10 }`  returns false

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes required :muscle:
Hi there @okikiola11 
Great job so far :+1:. Almost all methods are working as the original ones.
We need to make some changes before moving on. Please check my comments below.
You can use [this link](https://ruby-doc.org/core-2.7.1/Enumerable.html) as a guide

Kindly submit for code review again after making these changes
Feel free to message on Slack if you have any question about this review :smile: 
Happy coding :computer: 

**Sergio Torres**

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    result = []
+    my_each { |item| result.push(item) if yield(item) }
+    result
+  end
+
+  def my_all?(args = nil)
+    condition = true
+    if block_given?
+      my_each { |item| condition = false if yield(item) }
```
Comment (majovanilla) [inline enumerable_methods.rb]: This part is not working. Your method is returning false when given a truthy block of code.
`[1,2,3].my_all? { |num| num < 10 }`

I don't think you need to create a variable `condition` you can just return true or false. I would recommend returning false after you call my_each method (next line) if one element is not in your array.

Your method is returning TRUE when applying it to a falsy element.
`[1,2,3].my_all?(3)`

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    elsif args.nil?
+      my_each { |item| condition = false if item == true }
```
Comment (majovanilla) [inline enumerable_methods.rb]: This part is not working. Your method is returning false when applying it to a truthy element. 
`[1,2,3].my_all?`

File: enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    result = []
+    my_each { |item| result.push(item) if yield(item) }
+    result
+  end
+
+  def my_all?(args = nil)
```
Comment (majovanilla) [inline enumerable_methods.rb:30]: Your method is returning TRUE when applying it to a falsy element.
`[1,2,3, 'string', 'another', 'text'].my_all?(String)`

Your method is returning TRUE when applying it to a falsy element.
`['string', 'another', 'spacing'].my_all?(/ing/)`

File: enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    else
+      my_each { |item| condition = false if item === args }
+    end
+    condition
+  end
+
+  def my_any?(args = nil)
+    condition = false
+    if args.nil?
+      my_each { |item| condition = true if yield(item) }
```
Comment (majovanilla) [inline enumerable_methods.rb]: This part is not working. Your method is returning an error when no block is given.
`[1,2,3, 'string', 'another', 'text'].my_any?`

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey team, 👋

## You did a *great* job on this project! 💜 You are almost done with this!!

There are still some changes 👇 you need to implement to approve this request 🔍

- Check the comments below and make the necessary changes.

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @okikiola11      
#### Great Job so far , but there are a few more things required. 
### General
- [ ] `my_all?`  
- when no block or argument is given returns true when none of the collection members are false or nil:
 `p [1, true, 'hi', []].my_all?` should return `true`.
- when a class is passed as an argument returns true if all the collection is a member of such class
`p [1,1,1].my_all?(Integer)` should return `true`
-  when a Regex is passed as an argument returns true if all the collection matches the Regex
`p ['d','d'].my_all?(/d/)` should return `true`

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by shshamim63** [comment]
## Status :+1: Approved
___
Hi @okikiola11 :wave: ,
Awesome job :clap: All of your methods are working properly. Readme is descriptive, no explicit use of return, just found an issue regarding to your rubocop config. You should not change the default configuration of the rubocop config, But I can see you have added some ruleset to disable them. As the previous reviewer did not mention about it, but it is crucial so I am marking it as an optional change.
___
### Optional Change
___
- make sure that you are using the linter config mentioned in [here](https://github.com/microverseinc/linters-config/tree/master/ruby ). If you need to disable any rules do it using inline comment.
___
#### This :+1: is from me, you can go ahead.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Pull request merged**